File: codes/vimkim/3_longestsubstringwithoutrepeatingcharacters.go
package main

import "fmt"

func lengthOfLongestSubstring(s string) int {
    set := make(map[byte]bool)
    l, r, ans := 0, 0, 0
    N := len(s)

    for r < N {
        if !set[s[r]] {
            set[s[r]] = true
            r++
            if r - l > ans {
                ans = r - l
            }
        } else {
            delete(set, s[l])
            l++
        }
    }
    return ans
}

File: codes/jiheon/3_longestsubstringwithoutrepeatingcharacters.rs
use std::char;
use std::collections::HashSet;

impl Solution {
    pub fn length_of_longest_substring(s: String) -> i32 {
        let s: Vec<char> = s.chars().collect();
        let mut ans = 0;
        let mut left = 0usize;
        let mut set: HashSet<char> = HashSet::new();

        for (i, c) in s.iter().enumerate() {
            if !set.insert(*c) { // fn HashSet.insert(c) returns true if c was not present in set before inserting it.
                while s[left] != *c {
                    set.remove(&s[left]);
                    left += 1;
                }
                left += 1;
            } else {
                ans = ans.max(i - left + 1);
            }
        }

        ans as _
    }
}